pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,971 @@
+				let max_limit_with_slippage = estimated_amount_in
+					.checked_add(slippage_amount)
+					.ok_or(ArithmeticError::Overflow)?;
+
+				let max_limit = min(*max_limit, max_limit_with_slippage);
+				ensure!(amount_in <= max_limit, Error::<T>::TradeLimitReached);
+
+				pallet_route_executor::Pallet::<T>::buy(
+					origin,
+					*asset_in,
+					*asset_out,
+					(*amount_out).into(),
+					max_limit.into(),
+					route.to_vec(),
+				)
+			}
+		}
+	}
+
+	fn replan_or_complete(
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>,
+		current_blocknumber: T::BlockNumber,
+	) -> DispatchResult {
+		Self::deposit_event(Event::TradeExecuted {
+			id: schedule_id,
+			who: schedule.owner.clone(),
+		});
+
+		RetriesOnError::<T>::remove(schedule_id);
+
+		let remaining_amount_to_use: Balance =
+			RemainingAmounts::<T>::get(schedule_id).defensive_ok_or(Error::<T>::InvalidState)?;
+		let transaction_fee = Self::get_transaction_fee(&schedule.order)?;
+		if remaining_amount_to_use <= transaction_fee {
+			Self::complete_schedule(schedule_id, schedule);
+			return Ok(());
+		}
+
+		//In buy we complete with returning leftover, in sell we sell the leftover in the next trade
+		if let Order::Buy { amount_out, route, .. } = &schedule.order {
+			let amount_to_unreserve: Balance = Self::get_amount_in_for_buy(amount_out, route)?;
+
+			let amount_for_next_trade: Balance = amount_to_unreserve
+				.checked_add(transaction_fee)
+				.ok_or(ArithmeticError::Overflow)?;
+
+			if remaining_amount_to_use < amount_for_next_trade {
+				Self::complete_schedule(schedule_id, schedule);
+				return Ok(());
+			}
+		}
+
+		let next_execution_block = current_blocknumber
+			.checked_add(&schedule.period)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Self::plan_schedule_for_block(&schedule.owner, next_execution_block, schedule_id)?;
+
+		Ok(())
+	}
+
+	fn retry_schedule(
+		schedule_id: ScheduleId,
+		schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>,
+		current_blocknumber: T::BlockNumber,
+	) -> DispatchResult {
+		let number_of_retries = Self::retries_on_error(schedule_id);
+
+		ensure!(
+			number_of_retries < T::MaxNumberOfRetriesOnError::get(),
+			Error::<T>::MaxRetryReached
+		);
+
+		RetriesOnError::<T>::mutate(schedule_id, |retry| -> DispatchResult {
+			retry.saturating_inc();
+			Ok(())
+		})?;
+
+		let retry_multiplier = 2u32
+			.checked_pow(number_of_retries.into())
+			.ok_or(ArithmeticError::Overflow)?;
+		let retry_delay = SHORT_ORACLE_BLOCK_PERIOD
+			.checked_mul(retry_multiplier)
+			.ok_or(ArithmeticError::Overflow)?;
+		let next_execution_block = current_blocknumber
+			.checked_add(&retry_delay.into())
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Self::plan_schedule_for_block(&schedule.owner, next_execution_block, schedule_id)?;
+
+		Ok(())
+	}
+
+	fn price_change_is_bigger_than_max_allowed(schedule: &Schedule<T::AccountId, T::AssetId, T::BlockNumber>) -> bool {
```

Comment:
as predicate this should be really named `is_price_change_bigger_than_max_allowed`, but i would prefer simpler and more descriptive `is_price_unstable` with `PriceUnstable` error attached